### PR TITLE
frontend: Display and enforce password complexity on the change password screen

### PR DESCRIPTION
**Pull request by reivilibre**
Part of: #172 

Will need to follow up after this PR to do this on every site of the change password form (probably converting the password recovery form to React like we discussed earlier) but want to keep this minimal to review for now.

Also would be worth figuring out how we lazy load the zxcvbn library and dicts.

### CODE REVIEW

File: frontend/locales/en.json [hunk 1]
```diff
@@ -127,6 +127,51 @@
...
+      "placeholder": "Password strength",
+      "score": {
+        "0": "Extremely weak password",
+        "1": "Very weak password",
+        "2": "Weak password",
+        "3": "Strong password",
+        "4": "Very strong password"
+      },
+      "suggestion": {
+        "allUppercase": "Capitalize some, but not all letters.",
```
Comment (sandhose) [inline frontend/locales/en.json:140]: all the other keys use underscores, could we do that here as well?

File: frontend/src/utils/password_complexity/index.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      return t("frontend.password_strength.score.2");
+    case 3:
+      return t("frontend.password_strength.score.3");
+    case 4:
+      return t("frontend.password_strength.score.4");
+  }
+}
+
+/** Returns a translated string corresponding to a password improvement suggestion from zxcvbn-ts. */
+function translateSuggestion(suggestionCode: string): string | undefined {
```
[the remark this answers is not in the document]
Comment (sandhose) [reply frontend/src/utils/password_complexity/index.ts:122]: type of that function is `i18next.TFunction<"frontend", undefined>` I think?

File: frontend/src/utils/password_complexity/index.ts
[the remark this answers is not in the document]
Comment (reivilibre) [reply frontend/src/utils/password_complexity/index.ts:1]: yes please, I think I'd be a bit lost at trying to do that (probably)

File: frontend/src/routes/password.change.index.tsx [hunk 3]
```diff
@@ -213,9 +267,28 @@ function ChangePassword(): React.ReactNode {
...
+              onChange={(e) => setNewPassword(e.target.value)}
+            />
+
+            <Form.ErrorMessage
+              match={(_value, _) =>
+                passwordComplexity.score < minPasswordComplexity
+              }
+            >
+              {t("frontend.password_strength.too_weak")}
+            </Form.ErrorMessage>
```
[the remark this answers is not in the document]
Comment (reivilibre) [reply frontend/src/routes/password.change.index.tsx:279]: I tried this but this does 2 things wrong:

* shows the error message immediately before you typed a password
* doesn't prevent you from submitting the form

It's not very obvious is it... but not sure if there's a more idiomatic way to do it that actually preserves that? :S

[the remark this answers is not in the document]
Comment (sandhose) [reply frontend/src/routes/password.change.index.tsx:279]: Ah right! What's the behaviour if you do then something like `<ErrorMessage match="badInput" forceMatch={passwordComplexity.score < minPasswordComplexity}>`? That's supposed to be used when you have external validation or something?

File: frontend/src/routes/password.change.index.tsx [hunk 4]
```diff
@@ -86,7 +86,10 @@ export const Route = createFileRoute("/password/change/")({
   component: ChangePassword,
 });
 
-const usePasswordComplexity = (password: string): PasswordComplexity => {
+const usePasswordComplexity = (
+  password: string,
+  t: TFunction<"frontend", undefined>,
+): PasswordComplexity => {
```
Comment (sandhose) [inline frontend/src/routes/password.change.index.tsx:92]: As this one is already a hook, it's fine to `useTranslation` in the hook body here if you want, it won't cost more to call that multiple times for the same component

**Review by sandhose — APPROVED** [review]
Looks good! Will look at code splitting tomorrow :)

**Pull request merged**